compiler/internal/codegen/testdata/TestCodeGenMain__variants.golden

```diff
@@ -1505,10 +1517,145 @@ func DI(ctx context.Context) error {
 	}
 	return svc.DI(ctx)
 }
+
+
+// config unmarshallers for service svc
+package svc
+
+import (
+	"encoding/json"
+	etype "encore.app/__encore/etype"
+	config "encore.dev/config"
+	"fmt"
+	jsoniter "github.com/json-iterator/go"
+	"strconv"
+	"time"
+)
+
+/*
+These functions are automatically generated and maintained by Encore to allow config values
+to be unmarshalled into the correct types. They are not intended to be used directly. They
+are automatically updated by Encore whenever you change the data types used within your
+calls to config.Load[T]().
+*/
+
+func EncoreInternal_UnmarshalConfig_FooParams(itr *jsoniter.Iterator, path []string) (obj FooParams) {
+	itr.ReadObjectCB(func(itr *jsoniter.Iterator, field string) bool {
+		switch field {
+		case "Name":
+			obj.Name = itr.ReadString()
+		}
+		return true
+	})
+	return
+}
+
+func EncoreInternal_UnmarshalConfig_Cfg(itr *jsoniter.Iterator, path []string) (obj Cfg) {
+	itr.ReadObjectCB(func(itr *jsoniter.Iterator, field string) bool {
+		switch field {
+		case "DefaultFooParams":
+			obj.DefaultFooParams = config.CreateValue[FooParams](EncoreInternal_UnmarshalConfig_FooParams(itr, append(path, "DefaultFooParams")), append(path, "DefaultFooParams"))
+		case "ReadOnly":
+			obj.ReadOnly = config.CreateValue[bool](itr.ReadBool(), append(path, "ReadOnly"))
+		case "MagicNumber":
+			obj.MagicNumber = itr.ReadInt()
+		case "Sub":
+			obj.Sub = EncoreInternal_UnmarshalConfig_SubType(itr, append(path, "Sub"))
+		}
+		return true
+	})
+	return
+}
+
+func EncoreInternal_UnmarshalConfig_SubType(itr *jsoniter.Iterator, path []string) (obj SubType) {
+	itr.ReadObjectCB(func(itr *jsoniter.Iterator, field string) bool {
+		switch field {
+		case "FieldA":
+			obj.FieldA = config.CreateValue[string](itr.ReadString(), append(path, "FieldA"))
+		case "FieldB":
+			obj.FieldB = config.CreateValue[[]string](config.ReadArray[string](itr, func(itr *jsoniter.Iterator, idx int) string {
+				return itr.ReadString()
+			}), append(path, "FieldB"))
+		case "FieldC":
+			obj.FieldC = config.ReadArray[int](itr, func(itr *jsoniter.Iterator, idx int) int {
+				return config.CreateValue[int](itr.ReadInt(), append(path, strconv.Itoa(idx)))
+			})
+		case "FieldD":
+			obj.FieldD = config.ReadMap[int, string](itr, func(itr *jsoniter.Iterator, keyAsString string) (int, string) {
+				// Decode the map key from the JSON string to the underlying type it needs to be
+				keyDecoder := &etype.Marshaller{}
+				key := keyDecoder.ToInt("keyAsString", key, true)
+				if keyDecoder.LastError != nil {
+					panic(fmt.Sprintf("unable to decode the config: %v", keyDecoder.LastError))
+				}
+
+				return key, config.CreateValue[string](itr.ReadString(), append(path, keyAsString))
+			})
+		case "FieldE":
+			obj.FieldE = func() (obj struct {
+				SubA string
+				SubB FooParams
+				Blah time.Time
+				Foo  []byte
+				Json json.RawMessage
+			}) {
+				itr.ReadObjectCB(func(itr *jsoniter.Iterator, field string) bool {
```

Comment:
This gets a bit messy. Could we maybe extract a FieldE_Decoder func?

Reply:
It's not a named type, and I need to track the type arguments for generics, so for generic purposes this is actually cleaner (code no in this PR)